Fix outline: auto in user agent style sheets

nt1m opened the pull request:
#### a8ca03b6a0665f0137a21245513050ed69561893
<pre>
Fix outline: auto in user agent style sheets
<a href="https://bugs.webkit.org/show_bug.cgi?id=241102">https://bugs.webkit.org/show_bug.cgi?id=241102</a>
<a href="https://rdar.apple.com/94116168">rdar://94116168</a>

Reviewed by Darin Adler.

It turns out the focus ring width in RenderTheme does not match the
visible width on macOS, which leads to platformFocusRingOffset() being
too wide. Fix this and remove negative outline-offset hacks from user
agent style sheets. We also rename platformFocusRingOffset() to
platformFocusRingWidthOffset() to avoid confusion with outline-offset.

We change outline-width on focus in
imported/blink/transitions/transition-shape-outside-crash.html as
otherwise there is no transition anymore as html.css no longer does
that.

We also properly reflect Adwaita as having a width of 2 now. And while
this is correct and helps for some tests it also clashes with the
default of outline-width. Bug 310543 tracks this for Adwaita
maintainers.

In RenderImage we make the macOS code path the shared code path. Since
the user agent style sheet no longer sets the default color, this
preserves that behavior for image maps on other ports.

Test: LayoutTests/fast/forms/input-search-focusring.html
Canonical link: <a href="https://commits.webkit.org/309812@main">https://commits.webkit.org/309812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e6e341b854116d9acdb766e05b7a8cd6bc6a9c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/151844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/24625 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/168/builds/18195 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/160586 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/105301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/153718 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/25118 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/24922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/117280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/105301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/154804 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/19455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/136263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/97995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/18540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/156/builds/24922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk3-libwebrtc~~](https://ews-build.webkit.org/#/builders/173/builds/8421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/128172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/14166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/163050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 ios-safer-cpp~~](https://ews-build.webkit.org/#/builders/174/builds/6199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/169/builds/15758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/163050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/24424 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/24922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/163050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/24425 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/135962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/81000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/23309 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/20513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/12737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/24042 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/88327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/23733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/23893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/23794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (changes requested):
Seems like a good idea but not passing tests yet.

darinadler commented on Source/WebCore/rendering/ios/RenderThemeIOS.h:
```diff
@@ -138,6 +138,7 @@ class RenderThemeIOS final : public RenderThemeCocoa {
     Color platformActiveSelectionBackgroundColor(OptionSet<StyleColorOptions>) const final;
     Color platformInactiveSelectionBackgroundColor(OptionSet<StyleColorOptions>) const final;
     Color platformFocusRingColor(OptionSet<StyleColorOptions>) const final;
+    float platformFocusRingWidth() const final { return 3; }
```
Where does this number 3 come from? How was it determined? I think we need a comment, not just an unexplained constant. This was already true in the old code.

Is there a good reason for the number to be in the header rather than in the `.cpp` file?

pxlcoder replied:
I believe this is matching `+[UIFocusRingStyle borderThickness]`, which we use in `GraphicsContextCG::drawFocusRing`.

It would be nice to have `platformFocusRingWidth` and  `drawFocusRing` call into the same method.

darinadler commented on Source/WebCore/rendering/RenderTheme.h:
```diff
@@ -200,8 +200,8 @@ class RenderTheme {
     virtual Color platformFocusRingColor(OptionSet<StyleColorOptions>) const { return Color::black; }
-    static float platformFocusRingWidth() { return 3; }
-    static float platformFocusRingOffset(float outlineWidth) { return std::max<float>(outlineWidth - platformFocusRingWidth(), 0); }
+    virtual float platformFocusRingWidth() const { return 5; }
```
Where does this number 5 come from? How was it determined? I think we need a comment, not just an unexplained constant. This was already true in the old code.

Is there a good reason for the number to be in the header rather than in the `.cpp` file?

Why is this number 5 a new default for all themes other than iOS? The change log and bug title both talk about macOS, but we’re changing it here everywhere except for iOS. I presume there is a good reason for this, but I don’t know what it is.

Why does this need to be determined at runtime and use a virtual function? I don’t think I understand why some things in `RenderTheme.h` are simply constants or done with `#if` and others are based on virtual functions and derived classes at runtime. Do we actually need the same binary to support multiple themes? Seems like this is a leftover of an unusual past choice and it would be nice to move in the right direction.

nt1m replied:
Both of these values were previously in html.css, I just reflected those. Let me see if I can find out more about the rationale

pxlcoder replied:
I think macOS should also be 3, matching the value in AppKit. Sadly that constant isn't exposed like the UIKit one.

I believe the 5px from before was because of the outline-offset of -2px on macOS only, which is now being removed.

darinadler replied:
I have a dim memory of the reason for the -2px that might be helpful.

A long time ago, the way Mac focus rings were rendered, there was a part of the ring that visually didn’t look like it was part of the ring, but was a faint trailing off bit of it, kind of like a barely visible aura or light shadow. Conceptually the ring’s thickness was only 3px, but in terms of invalidation and repainting there was some effect outside the 3px, I think it extended 2px further. We needed the web engine to know about this so it could properly paint the rings when they appeared and disappeared without leaving behind residue, but it didn’t make sense to represent this as part of the outline thickness. At the time, with a less sophisticated understand of CSS than the rest of you, I thought it was important to both know where the visual extent of the outline ended (3px out) and where the complete rendering extent ended (2px further out). I seem to recall that the rectangle you would pass in needed to be outset by the 3px, but the actual drawing would go 2px further out.

I don’t know if this memory is correct, but if so, has this changed? If it has, when? If not, do we have a new solution to this old problem? Or another way of asking this is: What is the correct way to express this kind of thing in our CSS and rendering machinery?

nt1m replied:
I can see how this made sense with the slight 3D border we had pre-liquid glass, but with liquid glass, the border is no longer semi-transparent, so the focusring no longer shows underneath:

<img width="804" height="300" alt="image" src="https://github.com/user-attachments/assets/bd55866f-02f3-4e00-bdb4-d10073d8e8ae" />

I checked using the digital color meter, and the border-color is identical with or without the focusring.

I'll change the focusring to be 3px instead.

darinadler reviewed (approved):
I don’t understand why fixing macOS is best done by changing all platforms other than iOS.

darinadler commented on LayoutTests/TestExpectations:
```diff
@@ -632,8 +632,8 @@ imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-mode
 imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/customizable-select-multiple-popup/select-multiple-popup-appearance.tentative.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/customizable-select/focus-ring-rendering.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/customizable-select/select-accessibility-minimum-target-size.html [ Skip ]
+imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/customizable-select/select-appearance-active.html [ ImageOnlyFailure ]
```
Do we understand why this fails?

annevk replied:
This is no longer failing. Everything test-wise should now be a progression or preserving the status quo.

darinadler commented on Source/WebCore/rendering/RenderTheme.h:
```diff
@@ -200,8 +200,11 @@ class RenderTheme {
     virtual Color platformFocusRingColor(OptionSet<StyleColorOptions>) const { return Color::black; }
+
+    // On iOS, this matches [UIFocusRingStyle borderThickness] and on macOS this matches AppKit.
     static float platformFocusRingWidth() { return 3; }
```
Maybe should mark this `constexpr`?

nt1m committed:
Fix outline: auto in user agent style sheets
https://bugs.webkit.org/show_bug.cgi?id=241102
rdar://94116168

Reviewed by Darin Adler.

It turns out the focus ring width in RenderTheme does not match the
visible width on macOS, which leads to platformFocusRingOffset() being
too wide. Fix this and remove negative outline-offset hacks from user
agent style sheets. We also rename platformFocusRingOffset() to
platformFocusRingWidthOffset() to avoid confusion with outline-offset.

We change outline-width on focus in
imported/blink/transitions/transition-shape-outside-crash.html as
otherwise there is no transition anymore as html.css no longer does
that.

We also properly reflect Adwaita as having a width of 2 now. And while
this is correct and helps for some tests it also clashes with the
default of outline-width. Bug 310543 tracks this for Adwaita
maintainers.

In RenderImage we make the macOS code path the shared code path. Since
the user agent style sheet no longer sets the default color, this
preserves that behavior for image maps on other ports.

Test: LayoutTests/fast/forms/input-search-focusring.html
Canonical link: https://commits.webkit.org/309812@main
